src/probnum/kernels/_kernel.py

```diff
@@ -0,0 +1,302 @@
+
+        The dimension of inputs to the covariance function :math:`k : \\mathbb{R}^{
+        d_{in}} \\times \\mathbb{R}^{d_{in}} \\rightarrow
+        \\mathbb{R}^{d_{out} \\times d_{out}}`.
+        """
+        return self.__input_dim
+
+    @property
+    def output_dim(self) -> int:
+        """Dimension of the evaluated covariance function.
+
+        The resulting evaluated kernel :math:`k(x_0, x_1) \\in
+        \\mathbb{R}^{d_{out} \\times d_{out}}` has *shape=(output_dim,
+        output_dim)*.
+        """
+        return self.__output_dim
+
+    def _check_and_transform_input(
+        self,
+        x0: _InputType,
+        x1: Optional[_InputType] = None,
+    ) -> Tuple[np.ndarray, np.ndarray, bool]:
+        """Transform inputs to the kernel matrix.
+
+        Transforms inputs into :class:`numpy.ndarray` and standardizes their shape.
+
+        Parameters
+        ----------
+        x0 :
+            First input to the covariance function.
+        x1 :
+            Second input to the covariance function.
+
+        Returns
+        -------
+        x0 :
+            First input to the covariance function.
+        x1 :
+            Second input to the covariance function.
+        equal_inputs :
+            Are the two inputs the same?
+
+        Raises
+        -------
+        ValueError :
+            If input shapes of x0 and x1 do not match the kernel input dimension or
+            each other.
+        """
+        # pylint: disable="too-many-boolean-expressions"
+        # Transform into array and add second argument
+        x0 = np.asarray(x0)
+        equal_inputs = False
+        if x1 is None:
+            x1 = x0
+            equal_inputs = True
+        else:
+            x1 = np.asarray(x1)
+
+        # Error message
+        err_msg = (
+            f"Argument shapes x0.shape={x0.shape} and x1.shape="
+            f"{x1.shape} do not match kernel input dimension "
+            f"{self.input_dim}. Try passing either two vectors or two matrices with the "
+            f"second dimension equalling the kernel input dimension."
+        )
+
+        # Check and promote shapes
+        if x1 is None:
+
+            if (
+                (x0.ndim == 0 and self.input_dim > 1)  # Scalar input
+                or (x0.ndim == 1 and x0.shape[0] != self.input_dim)  # Vector input
+                or (x0.ndim >= 2 and x0.shape[1] != self.input_dim)  # Matrix input
+            ):
+                raise ValueError(err_msg)
+            equal_inputs = True
+        else:
+            # Promote unequal shapes
+            if x0.ndim < 2 and x1.ndim == 2:
+                x0 = np.atleast_2d(x0)
+            if x1.ndim < 2 and x0.ndim == 2:
+                x1 = np.atleast_2d(x1)
+            if x0.ndim != x1.ndim:  # Shape mismatch
+                raise ValueError(err_msg)
+
+            # Check shapes
+            if (
+                (x0.ndim == 0 and self.input_dim > 1)  # Scalar input
```

Comment:
how about `def is_scalar_input(x0): return x0.ndim == 0 and self.input_dim > 1)` and similar functions in line 183 or so?